pyccl/pk2d.py

```diff
@@ -130,17 +130,34 @@ def __init__(self, pkfunc=None, a_arr=None, lk_arr=None, pk_arr=None,
                                                         int(is_logp), status)
         check(status, cosmo=cosmo)
 
+    def __eq__(self, other):
+        # Check the object class.
+        if self.__class__ is not other.__class__:
+            return False
+        # If the objects contain no data, return early.
+        if not (self or other):
+            return True
+        # Check extrapolation orders.
+        if not (self.extrap_order_lok == other.extrap_order_lok
+                and self.extrap_order_hik == other.extrap_order_hik):
+            return False
+        # Check the individual splines.
+        a1, lk1, pk1 = self.get_spline_arrays()
+        a2, lk2, pk2 = other.get_spline_arrays()
+        return ((a1 == a2).all() and (lk1 == lk2).all()
+                and np.allclose(pk1, pk2, atol=0, rtol=1e-12))
```

Comment:
For this particular line of code we can define a `ccl.spline_params` parameter that sets `rtol`. I chose 1e-12 because when I tested it `np.max(np.abs( 1 - pk / (2*pk/2) ))` was about `5e-13` due to truncation errors, which means that with what we currently have in `master`, `pk == 2*pk/2` will more often than not return `False`.

Let me know what you think about the additional spline parameter.

Reply:
I'd rather we check pk1 == pk2 to avoid any surprises at all. This is what anyone would expect __eq__ to do. No reason to complicate things.

Reply:
The problem is that due to machine precision, if we use `rtol=1e-12` this check will hardly ever pass, so all this eq-checking goes to waste. We want eq to return True if the pk will give us the same theoretical prediction, and having an rtol of 1e-12 will.
That question was more on if we are going to put it somewhere as a global tolerance parameter, or if we'll leave it hard-coded like that. I'm more inclined to do the first but not sure.

Reply:
Note, that if we implement a global control like we have for `FancyRepr` at the moment, all these checks can be disabled so we can use Python's default id-checking.

Reply:
Also, following the logic in previous comments, split this into three ifs for each of the arrays, so you don't necessarily need to go through all of them if e.g. the first one fails.

Reply:
I agree this will only pass when both pk2ds have been initialised with exactly the same arrays, but that's the intended use of eq. Compare them directly.

Reply:
1. No need to split them because Python will exit early if one of them is `False` in the chain. For example `True and print("hi")` prints hi, but `False and print("hi")` doesn't print anything.
2. No, eq shows us equivalence, not absolute equality. It is up to us developers to decide whether we want two arrays that are 1e-12 off to be equivalent. And I recall this was suggested in the past in one of the issues, but can't remember where exactly.

Reply:
Like I said, at the moment, `pk == 2*pk/2` is `False`, so we want to avoid that.

Reply:
Okay, I'm fine to use `np.array_equal` but doesn't it strike bad when you realize that `pk != 2*pk/2`? If not, I'll do it.

Reply:
Yep, array_equal is fine. I think the fact that pk != 2*pk/2 makes perfect sense :-)

Reply:
Ok final question. How would it sound to you if we indeed made `rtol` a global var but set it by default to zero so it exactly emulates the behaviour of `np.array_equal`? I am making these changes now.

Reply:
done, I set a glob var in `parameters.py` because it was the easiest thing to do, rather than change everything to `array_equal`.

Reply:
This just adds more complexity to the code that we don'tneed. There aren't so many places where you need to make this change (2?)

Reply:
done